nvflare/private/fed/cmi.py

```diff
@@ -0,0 +1,248 @@
+            message.set_header(self.HEADER_KEY_PEER_PROPS, public_props)
+            ssid = fl_ctx.get_prop(CellMessageHeaderKeys.SSID)
+            if ssid:
+                message.set_header(self.HEADER_SSID, ssid)
+            project_name = fl_ctx.get_prop(CellMessageHeaderKeys.PROJECT_NAME)
+            if project_name:
+                message.set_header(self.HEADER_PROJECT_NAME, project_name)
+            client_name = fl_ctx.get_prop(FLContextKey.CLIENT_NAME)
+            if client_name:
+                message.set_header(self.HEADER_CLIENT_NAME, client_name)
+            client_token = fl_ctx.get_prop(CellMessageHeaderKeys.TOKEN)
+            if client_token:
+                message.set_header(self.HEADER_CLIENT_TOKEN, client_token)
+
+    @staticmethod
+    def _make_peer_ctx(props: dict) -> FLContext:
+        ctx = FLContext()
+        ctx.set_public_props(props)
+        return ctx
+
+    @staticmethod
+    def _convert_return_code(rc: CellReturnCode):
+        return CellMessageInterface.RC_TABLE.get(rc, ReturnCode.ERROR)
+
+    @abstractmethod
+    def send_to_cell(
+        self,
+        targets: [],
+        channel: str,
+        topic: str,
+        request: Shareable,
+        timeout: float,
+        fl_ctx: FLContext,
+        bulk_send=False,
+    ) -> dict:
+        pass
+
+
+class JobCellMessenger(CellMessageInterface):
+    def __init__(self, engine, job_id: str):
+        super().__init__(engine)
+
+        self.job_id = job_id
+
+        self.cell.add_incoming_request_filter(channel="*", topic="*", cb=self._filter_incoming)
+        self.cell.add_incoming_reply_filter(channel="*", topic="*", cb=self._filter_incoming)
+        self.cell.add_outgoing_request_filter(channel="*", topic="*", cb=self._filter_outgoing)
+        self.cell.add_outgoing_reply_filter(channel="*", topic="*", cb=self._filter_outgoing)
+
+    def _filter_incoming(self, message: Message):
+        job_id = message.get_header(self.HEADER_JOB_ID)
+        if job_id and job_id != self.job_id:
+            self.logger.error(f"received job id {job_id} != my job id {self.job_id}")
+
+    def _filter_outgoing(self, message: Message):
+        message.set_header(self.HEADER_JOB_ID, self.job_id)
+
+    def send_to_cell(
+        self,
+        targets: [],
+        channel: str,
+        topic: str,
+        request: Shareable,
+        timeout: float,
+        fl_ctx: FLContext,
+        bulk_send=False,
+    ) -> dict:
+        """Send request to the job cells of other target sites.
+        Args:
+            targets (list): list of client names that the request will be sent to
+            channel (str): channel of the request
+            topic (str): topic of the request
+            request (Shareable): request
+            timeout (float): how long to wait for result. 0 means fire-and-forget
+            fl_ctx (FLContext): the FL context
+            bulk_send: whether to bulk send this request
```

Comment:
this `bulk_send` arg is only checked if `timeout` is 0.
So maybe we improve the docstring to say this is only checked in fire-and-forget, on the other note only checked in non-blocking situation.

Reply:
Added docstring explanation.